Libraries/LLM/Lora.swift

```diff
@@ -0,0 +1,356 @@
+        var fusedLinear = Linear(inputDimensions, outputDimensions, bias: linear.bias != nil)
+
+        let loraB = (scale * loraB.T).asType(dtype)
+        let loraA = loraA.T.asType(dtype)
+        var parameters = ModuleParameters()
+        parameters["weight"] = .value(weight + matmul(loraB, loraA))
+        if let bias = linear.bias {
+            parameters["bias"] = .value(bias)
+        }
+        fusedLinear.update(parameters: parameters)
+
+        if !deQuantize, let q = linear as? QuantizedLinear {
+            fusedLinear = QuantizedLinear.from(
+                linear: fusedLinear, groupSize: q.groupSize, bits: q.bits)
+        }
+
+        return fusedLinear
+    }
+
+    public func callAsFunction(_ x: MLXArray) -> MLXArray {
+        let y = linear(x.asType(self.dtype))
+        let z = matmul(matmul(x, self.loraA), self.loraB)
+        return y + scale * z
+    }
+}
+
+public struct LoRAParameters {
+    public var batchSize = 4
+    public var iterations = 1000
+
+    public var stepsPerReport = 10
+
+    public var stepsPerEval = 100
+    public var validationBatches = 10
+
+    public var saveEvery = 100
+    public var adapterURL: URL?
+
+    public init() {
+    }
+}
+
+/// Equivalent to `lora.py/iterate_batches()`
+struct LoRABatchIterator: Sequence, IteratorProtocol {
+
+    let dataset: [String]
+    let batchSize: Int
+    let tokenizer: Tokenizer
+
+    let train: Bool
+
+    var indices: [Int]
+    var index = 0
+
+    public init(dataset: [String], tokenizer: Tokenizer, batchSize: Int, train: Bool) {
+        self.dataset = dataset
+        self.batchSize = batchSize
+        self.tokenizer = tokenizer
+        self.train = train
+
+        self.indices = Array(0 ..< dataset.count)
+        if train {
+            indices.shuffle()
+        }
+    }
+
+    mutating public func next() -> (MLXArray, MLXArray, MLXArray)? {
+        if index >= indices.count {
+            if !train {
+                return nil
+            }
+
+            indices.shuffle()
+            index = 0
+        }
+
+        let endIndex = Swift.min(index + batchSize, indices.count)
+
+        let batch = (index ..< endIndex)
+            .map { tokenizer.encode(text: dataset[indices[$0]]) }
+        let lengths = batch.map { $0.count }
+        let maxLength = lengths.max() ?? 0
+
+        if maxLength > 2048 {
+            print(
+                """
+                [WARNING] Some sequences are longer than 2048 tokens.
+                Consider pre-splitting your data to save memory.
+                """)
+        }
+
+        // pad to the max length
+        let batchArray = MLXArray.zeros([lengths.count, maxLength], type: Int32.self)
+        for (j, (b, l)) in zip(batch, lengths).enumerated() {
+            batchArray[j, 0 ..< l] = MLXArray(b)
+        }
+
+        index = endIndex
+
+        return (batchArray[0..., .stride(to: -1)], batchArray[0..., 1...], MLXArray(lengths))
+    }
+
+}
+
+public enum LoRATrain {
```

Comment:
This is the public API -- just free functions grouped into this LoRATrain namespace.